### ISSUE TITLE
StateMechanic 1.0.3 is not compatible with Xamarin.iOS anymore

### ISSUE THREAD

**Issue by kaolick** [canton7/StateMechanic#4]
After upgrading StateMechanic to 1.0.3 in our iOS-App we get the following error when firing a state transition:

```
**System.ExecutionEngineException** has been thrown
Attempting to JiT compile method 'StateMechanic.StateMachine`1<StateMechanic.State>:RequestEver(StateMechanic.EventTransitionInvoker`1<StateMechanic.State>)' while running with --aot-only. See http://docs.xamarin.com/ios/about/limitations for more information.
```

The error only occurs on real devices, not on the simulator.

Here are the first few lines from the stack trace:

```
at StateMechanic.StateMachine`1[TState].StateMechanic.IEventDelegate.RequestEventFireFromEvent (StateMechanic.Event event, EventFireMethod eventFireMethod) [0x0000a] in <filename unknown>:0 
  at StateMechanic.Event.RequestEventFireFromEvent (EventFireMethod eventFireMethod) [0x0000f] in <filename unknown>:0 
  at StateMechanic.Event.Fire () [0x00000] in <filename unknown>:0 
```

**Comment by canton7** [comment]
Weird: I'm fairly sure that method hasn't changed since 1.0.2, but I won't have access to a computer to check for sure until later in the week. I'm also pretty sure that we don't do anything that the AOT doesn't support. I suspect this will probably end up going to the Xamarin team, but I'll have a closer look later on the week first.

**Comment by canton7** [comment]
Does this happen when firing any event, or only with particular events? If I were to create a minimal state machine with a single state, would that reproduce the issue?

I'm afraid I don't have an iOS device, so I might be relying on you until I can get a repro.

**Comment by kaolick** [comment]
I don't know (yet). I will create a minimum example asap and let you know.

**Comment by canton7** [comment]
I spoke to one of the mono/Xamarin guys, and he said it was probably related to that call to `RequestEventFireFromEvent<TTransitionInvoker>`, with the AOT not managing to figure out that it was called with an `EventTransitionInvoker` (for some reason), and could well be a bug.

The first step, of course, is to get a minimal repro. I'm going to try and repro it using `mono --full-aot` (I had a bit of a try and didn't get anywhere, but will try some more). 

A workaround might be to make `EventTransitionInvoker` a class. That comes at the cost of an allocation per event invocation (which I'd more or less got down to zero allocations for the happy case), but that's not the end of the world.

**Comment by kaolick** [comment]
I've implemented a minimum example where I fire a simple state transition on a button click. This causes the error I mentioned above. Here is the sample code of my state machine:

```
using StateMechanic;

namespace StateMechanicDemo.Shared
{
    public class DemoStateMachine : StateMachine<State>
    {
        State A;
        State B;
        Event AtoB;

        public DemoStateMachine()
        {
            A = CreateInitialState("A");
            B = CreateState("B");

            AtoB = new Event("A to B");

            A.TransitionOn(AtoB).To(B);
        }

        public void MakeTransitionFromAtoB()
        {
            AtoB.Fire();
        }
    }
}
```

**Comment by canton7** [comment]
OK, thanks for checking. I should have some time tomorrow to try and reproduce with mono and AOT.

**Comment by canton7** [comment]
Right! Managed to reproduce it under "pure" mono, using `--full-aot`.

**Comment by canton7** [comment]
Looks like this repros it:

```csharp
public static class Program
{
    public static void Main(string[] args)
    {
        var sm2 = new TestClass<int>(); // <-- Doesn't seem to matter what the parameter is
        sm2.Part1();
    }
}

public class TestClass<T> // <-- Needs to be a generic type
{
    public void Part1() // <-- Needs to be in this class - can't be in Program
    {
        var foo = new Foo();
        this.Part2(foo);
    }

    public void Part2<TFoo>(TFoo foo) where TFoo : IFoo // <-- Needs to be generic
    {
        Debug.WriteLine(foo.Bar); // <-- Needs to reference a member from foo
    }
}

public enum SomeEnum { One }

public interface IFoo
{
    SomeEnum Bar { get; } // <-- Needs to be an enum (an int/object/custom struct will work)
}

public struct Foo : IFoo
{
    public SomeEnum Bar { get; }
}
```

Compile with `--aot=full`, run with `--full-aot`.

**Comment by canton7** [comment]
Reported as https://bugzilla.xamarin.com/show_bug.cgi?id=52600. Let's see what happens...

I'll wait a couple of days to see what sort of response it gets, then consider trying to work around it.

**Comment by kaolick** [comment]
Thank you so far. It seems that some of your changes between 1.0.2 and 1.0.3 lead to this problem. I'd like to ask for a favour in the meantime: You made a bugfix regarding the order of OnEntry()-calls of StateGroup (https://github.com/canton7/StateMechanic/commit/e55a3e02c0379085de16eeea9cfc3fcf82ddfb54). It would be really great if you could provide a version only containing this bugfix. Thank you in advance.

**Comment by canton7** [comment]
I don't have a fix for this yet. I don't know what's causing it exactly. As I said, I'd like to give the mono guys a chance to respond, in case they come back with "oh it's this specific combination of things which triggers the bug", in which case I'll know what to do to work around it. If that doesn't happen, I'll start the process of trying different changes until it starts working again.

**Comment by kaolick** [comment]
Version 1.0.2 was working just fine. So you could just go back to this version, add the bugfix for StateGroup and deal with the other bug later.

**Comment by kaolick** [comment]
By the way, did you take a look at this: https://developer.xamarin.com/guides/ios/advanced_topics/limitations/

I guess it will have something to do with generics and/or dynamic code generation.

**Comment by canton7** [comment]
Yeah, that's the first thing I looked at. I think it is to do with calling that generic method with a struct parameter, but it isn't as simple as that: the AOT should be able to determine that it's called with a struct parameter, and indeed it does most of the time: the thing that breaks it is reading an *enum* member from the struct. This is all detailed in my code snippet above, and in the Mono issue I linked.

**Comment by canton7** [comment]
Can you try the intermediate build 1.0.3.124 please? You can download it from the NuGet feed below, or directly:

 - NuGet feed: https://ci.appveyor.com/nuget/statemechanic
 - Direct download: https://ci.appveyor.com/api/buildjobs/dlotby6laoim932e/artifacts/StateMechanic.1.0.3.124.nupkg

Thanks!

**Comment by kaolick** [comment]
Thank you. I'll have a look at it tomorrow.

**Comment by kaolick** [comment]
Works like a charm. Does this version include your StateGroup fix?

**Comment by canton7** [comment]
Awesome! Yeah, it's a modification on top of the current master. It's what I wanted to do: keep what I've introduced since 1.0.2, but work around the Mono issue.

I'll try and get that released today.

**Comment by canton7** [comment]
Right, that's released as 1.0.4.
